Add `String<A>` type with custom allocator parameter

shua opened the pull request:
<!-- homu-ignore:start -->
*[View all comments](https://triagebot.infra.rust-lang.org/gh-comments/rust-lang/rust/pull/149328)*
<!-- homu-ignore:end -->

This change is part of the `allocator_api` feature set [#32838](https://github.com/rust-lang/rust/issues/32838) (also see [wg-allocators roadmap] or [libs-team ACP]).
The previous attempts at adding an allocator parameter to string are at [rust-lang/rust#101551], and [rust-lang/rust#79500] (I think those authors should get much of the credit here, I am re-writing what they worked out in those threads).

## workaround

There is a type inference ambiguity introduced when adding a generic parameter to a type which previously had none, even when that parameter has a default value (more details in [rust-lang/rust#98931]). I've done the same [workaround] as [rust-lang/rust#101551], which is to make `alloc::string::String` a type alias to `String<Global>`, but I've arranged the modules a bit differently to make rebase/merges a bit easier.

This workaround unfortunately changes the type name of the `String` language item, and that would be user-facing in error or diagnostic output. I understand from [this comment](https://github.com/rust-lang/rust/pull/101551#issuecomment-2605604862) that this change is acceptable.

## changes to existing API

Most of the methods on the original `String` have been implemented for the generic version instead. I don't foresee anything more moving from `String<Global>` to `String<A>`, as the remaining methods are all constructors which implicitly use the `Global` allocator.

There are three general types of changes:
1. methods which don't allocate: here we just change `impl Foo for String` to `impl<A: Allocator> Foo for String<A>`
2. converting to/from other allocator generic types like `Vec<u8, A>` and `Box<str, A>`: here we can use the existing allocator from those types.
3. methods which clone from some other type with an allocator: here it's ambiguous whether the end result should be `String<A>`, `String<Global>`, or `String<impl Allocator + Default>`, etc; in general I try to leave these out of this change, but where some analogous change was made to `Vec<T, A>` I follow that.

## new methods

Some methods have been added to `String<A>` which are not strictly necessary to land this change, but are considered helpful enough to users, and close enough to existing precedent in `Vec<T, A>`. Specifically, 4 new constructors (`new_in`, `with_capacity_in`, `try_with_capacity_in`, `from_raw_parts_in`), 1 destructor (`into_raw_parts_with_alloc`), and 1 getter (`allocator`). Technically, the updated `from_utf8_unchecked` should be sufficient for constructing, but we can add some safe constructors so users don't have to sully themselves.

## not implemented

Variants of `from_utf{8,16}*` which internally allocate or use `Cow` have been punted on this PR, maybe a followup would make sense to either rewrite them, or add some `*_in` variant.

[wg-allocators roadmap]: https://github.com/rust-lang/wg-allocators/issues/7
[libs-team ACP]: https://github.com/rust-lang/libs-team/issues/101
[workaround]: https://github.com/rust-lang/rust/issues/79499#issuecomment-1238578527
[rust-lang/rust#101551]: https://github.com/rust-lang/rust/pull/101551
[rust-lang/rust#79500]: https://github.com/rust-lang/rust/pull/79500
[rust-lang/rust#98931]: https://github.com/rust-lang/rust/issues/98931

shua committed:
Add allocator param to String

This is a third attempt at implementing adding Allocator support to
the std lib's `String`.  Still stuck on the same issue with type
inference failing on the newly generic `String<A>`, but I opted to do
even less than the previous WIP work, and have added no new functions
(`String<A>` can be constructed via `Vec<u8, A>` or `Box<str, A>`),
and have moved only the struct definition to its own mod to make
rebasing a bit easier if/when main changes underneath me.

shua committed:
bless ui and mir-opt tests

shua committed:
update name in tests: ui, mir-opt, std, llvm cg

shua committed:
update debugger scripts

shua committed:
update docs references, fix linkchecker and tests

added some code in linkchecker to check the generic::String docs when
trying to resolve links to alloc::string::String type alias. There's
some lazy-loading that the browser does, but linkchecker doesn't, so
maybe some general-purpose solution would be better, but this seemed
better than a big list of exceptions.

shua committed:
add ctor/dtors: new_in, from_raw_parts_in, etc

Yes, you could just use `unsafe { from_utf8_unchecked }``, but people get
antsy about `unsafe`, so add some Vec ctor/dtor equivalents.

shua committed:
update unimpl diagnostics, clippy lints

This adds a diagnostic name `string_in_global` so that clippy can easily
check for the alloc::string::String type alias.

rust-log-analyzer commented:
The job **`aarch64-gnu-llvm-20-2`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/19684606441/job/56386988387) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/56386988387) [(plain)](https://github.com/rust-lang/rust/commit/29100c18b2baa53531ca4d6acd3b4d0e49d81eab/checks/56386988387/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
##[endgroup]
[TIMING:end] tool::ToolBuild { build_compiler: Compiler { stage: 0, host: aarch64-unknown-linux-gnu, forced_compiler: false }, target: aarch64-unknown-linux-gnu, tool: "linkchecker", path: "src/tools/linkchecker", mode: ToolBootstrap, source_type: InTree, extra_features: [], allow_features: "", cargo_args: [], artifact_kind: Binary } -- 0.142
[TIMING:end] tool::Linkchecker { compiler: Compiler { stage: 0, host: aarch64-unknown-linux-gnu, forced_compiler: false }, target: aarch64-unknown-linux-gnu } -- 0.000
##[group]Testing stage1 Linkcheck (aarch64-unknown-linux-gnu)
std/string/generic/struct.String.html:734: broken link - `std/std/primitive.str.html`
alloc/string/generic/struct.String.html:730: broken link - `alloc/std/primitive.str.html`
book/print.html:1063: broken link - `std/string/struct.String.html`
book/ch02-00-guessing-game-tutorial.html:367: broken link - `std/string/struct.String.html`
NOTE: if you are adding or renaming a markdown file in a mdBook, don't forget to register the page in SUMMARY.md
checked links in: 21.9s
number of HTML files scanned: 46371
number of HTML redirects found: 14531
number of links checked: 2562006
number of links ignored due to external: 123104
---
Command `/checkout/obj/build/aarch64-unknown-linux-gnu/stage1-tools-bin/linkchecker /checkout/obj/build/aarch64-unknown-linux-gnu/doc` failed with exit code 1
Created at: src/bootstrap/src/core/build_steps/tool.rs:1599:23
Executed at: src/bootstrap/src/core/build_steps/test.rs:170:77

Command has failed. Rerun with -v to see more details.
Bootstrap failed while executing `--stage 2 test --skip tests --skip coverage-map --skip coverage-run --skip library --skip tidyselftest`
Build completed unsuccessfully in 0:37:54
  local time: Tue Nov 25 22:10:30 UTC 2025
  network time: Tue, 25 Nov 2025 22:10:30 GMT
##[error]Process completed with exit code 1.
```

</details>

rust-log-analyzer commented:
The job **`aarch64-gnu-llvm-20-2`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/19695389560/job/56419332032) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/56419332032) [(plain)](https://github.com/rust-lang/rust/commit/87e91e6a1aec2f4c86081a2d2c0d7f3f3a0e7f90/checks/56419332032/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
##[endgroup]
[TIMING:end] tool::ToolBuild { build_compiler: Compiler { stage: 0, host: aarch64-unknown-linux-gnu, forced_compiler: false }, target: aarch64-unknown-linux-gnu, tool: "linkchecker", path: "src/tools/linkchecker", mode: ToolBootstrap, source_type: InTree, extra_features: [], allow_features: "", cargo_args: [], artifact_kind: Binary } -- 0.147
[TIMING:end] tool::Linkchecker { compiler: Compiler { stage: 0, host: aarch64-unknown-linux-gnu, forced_compiler: false }, target: aarch64-unknown-linux-gnu } -- 0.000
##[group]Testing stage1 Linkcheck (aarch64-unknown-linux-gnu)
book/print.html:1063: broken link - `std/string/struct.String.html`
book/ch02-00-guessing-game-tutorial.html:367: broken link - `std/string/struct.String.html`
checked links in: 21.7s
number of HTML files scanned: 46371
number of HTML redirects found: 14531
number of links checked: 2562010
number of links ignored due to external: 123104
number of links ignored due to exceptions: 24
number of intra doc links ignored: 9
errors found: 2
found some broken links
NOTE: if you are adding or renaming a markdown file in a mdBook, don't forget to register the page in SUMMARY.md
Command `/checkout/obj/build/aarch64-unknown-linux-gnu/stage1-tools-bin/linkchecker /checkout/obj/build/aarch64-unknown-linux-gnu/doc` failed with exit code 1
Created at: src/bootstrap/src/core/build_steps/tool.rs:1599:23
Executed at: src/bootstrap/src/core/build_steps/test.rs:170:77

Command has failed. Rerun with -v to see more details.
Bootstrap failed while executing `--stage 2 test --skip tests --skip coverage-map --skip coverage-run --skip library --skip tidyselftest`
Build completed unsuccessfully in 0:37:56
  local time: Wed Nov 26 07:48:06 UTC 2025
  network time: Wed, 26 Nov 2025 07:48:06 GMT
##[error]Process completed with exit code 1.
```

</details>

bors commented:
:umbrella: The latest upstream changes (presumably #147799) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

bors commented:
:umbrella: The latest upstream changes (presumably #150546) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

reddevilmidzy commented:
@rustbot author

shua commented:
@rustbot ready

Amanieu commented:
Could you run @bors try once you have rebased?

@bors delegate+

shua commented:
ah thanks, I didn't know if I should wait for CI jobs to pass before triggering bors (this is my first PR to rust)

Amanieu commented:
@craterbot check

shua commented:
for some reason, when rustdoc is building the reference docs, it thinks `alloc::string::String` is a struct, and so makes and ```[`String`]``` links into a link to `alloc/string/struct.String.html`. The correct file should be `type.String.html`, but I am not able to change the reference documentation because it's a git subtree. Should I add some exception to linkchecker, even if it is correctly flagging a bad link and followup with a PR to the reference docs, or should I spend more time trying to figure out what rustdoc is doing?

zachs18 commented:
> All stable functions that construct a String still force the global allocator to be used, right? That should be enough to guide type inference I would think when combined with a default for the type parameter for when you explicitly write String as type.

Note that using `String` as the base of a path does not apply the type parameter defaults, unlike using `<String>` as the base of the path, or using `String` as a type annotation. E.g. if you use `String::whatever` to get an associated item from a blanket-impl'd trait, that would no longer work since `String::path` would not constrain `A` (only `<String>::path` would): [playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2024&gist=61be86cf8aeb2db731ba09b2bae1b3d2).

See also my comments from my attempt at doing this: https://github.com/rust-lang/rust/pull/101551#issuecomment-1368602576 https://github.com/rust-lang/rust/pull/101551#issuecomment-2571281937

> [...] As to why adding the allocator parameter to Vec and other containers didn't break existing code, I don't recall all of the reasoning at the moment, but IIRC this issue doesn't affect Vec or other stdlib containers (as much) because String is the only one that doesn't *already* have any generics, so the other containers already have to rely on either explicit annotations or type inference for their element type(s), so adding the allocator type doesn't change much for them. [...]

rust-log-analyzer commented:
The job **`tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/25612267118/job/75184399315) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/75184399315) [(plain)](https://github.com/rust-lang/rust/commit/3be4d85cdd8c37e7a078d61c8e8c00be4e9ad853/checks/75184399315/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
[TIMING:end] tool::ToolBuild { build_compiler: Compiler { stage: 0, host: x86_64-unknown-linux-gnu, forced_compiler: false }, target: x86_64-unknown-linux-gnu, tool: "tidy", path: "src/tools/tidy", mode: ToolBootstrap, source_type: InTree, extra_features: [], allow_features: "", cargo_args: [], artifact_kind: Binary } -- 10.570
[TIMING:end] tool::Tidy { compiler: Compiler { stage: 0, host: x86_64-unknown-linux-gnu, forced_compiler: false }, target: x86_64-unknown-linux-gnu } -- 0.000
fmt check
Diff in /checkout/library/alloc/src/string.rs:347:
 /// [`Deref`]: core::ops::Deref "ops::Deref"
 /// [`as_str()`]: String::as_str
 #[cfg(doc)]
-pub struct String<A: Allocator = Global>{
+pub struct String<A: Allocator = Global> {
     vec: Vec<u8, A>,
 }
 
Diff in /checkout/library/alloc/src/string.rs:389:
     }
 }
 
-
 /// A possible error value when converting a `String` from a UTF-8 byte vector.
 ///
 /// This type is the error type for the [`from_utf8`] method on [`String`]. It
Diff in /checkout/library/alloc/src/string.rs:2243:
     /// assert_eq!(s, "Test Results: ✅❌❌");
     /// ```
     ///
-    #[cfg_attr(not(doc), doc = "[replacen]: ../../../std/primitive.str.html#method.replacen" )]
-    #[cfg_attr(doc, doc = "[replacen]: ../../std/primitive.str.html#method.replacen" )]
+    #[cfg_attr(not(doc), doc = "[replacen]: ../../../std/primitive.str.html#method.replacen")]
+    #[cfg_attr(doc, doc = "[replacen]: ../../std/primitive.str.html#method.replacen")]
     #[cfg(not(no_global_oom_handling))]
     #[unstable(feature = "string_replace_in_place", issue = "147949")]
     pub fn replace_first<P: Pattern>(&mut self, from: P, to: &str) {
fmt: checked 6852 files
Bootstrap failed while executing `test src/tools/tidy tidyselftest --extra-checks=py,cpp,js,spellcheck`
Build completed unsuccessfully in 0:00:37
  local time: Sat May  9 21:33:44 UTC 2026
  network time: Sat, 09 May 2026 21:33:44 GMT
```

</details>

rust-log-analyzer commented:
The job **`tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/25612481660/job/75185001590) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/75185001590) [(plain)](https://github.com/rust-lang/rust/commit/35843d3a0e55e1203468d8cb41d51cb6bf321bb7/checks/75185001590/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
     }
 }
 
-
 /// A possible error value when converting a `String` from a UTF-8 byte vector.
 ///
 /// This type is the error type for the [`from_utf8`] method on [`String`]. It
fmt: checked 6852 files
Bootstrap failed while executing `test src/tools/tidy tidyselftest --extra-checks=py,cpp,js,spellcheck`
Build completed unsuccessfully in 0:00:45
  local time: Sat May  9 21:45:12 UTC 2026
  network time: Sat, 09 May 2026 21:45:13 GMT
```

</details>

rust-log-analyzer commented:
The job **`aarch64-gnu-llvm-21-2`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/25612620307/job/75185359852) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/75185359852) [(plain)](https://github.com/rust-lang/rust/commit/8b23907d770876138e90cbf22e8d18dab9185108/checks/75185359852/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
##[endgroup]
Executing "/scripts/stage_2_test_set2.sh"
+ /scripts/stage_2_test_set2.sh
+ '[' 1 == 1 ']'
+ echo 'PR_CI_JOB set; skipping tidy'
+ SKIP_TIDY='--skip tidy'
+ ../x.py --stage 2 test --skip tidy --skip tests --skip coverage-map --skip coverage-run --skip library --skip tidyselftest
PR_CI_JOB set; skipping tidy
##[group]Building bootstrap
    Finished `dev` profile [unoptimized] target(s) in 0.04s
##[endgroup]
---
##[endgroup]
[TIMING:end] tool::ToolBuild { build_compiler: Compiler { stage: 0, host: aarch64-unknown-linux-gnu, forced_compiler: false }, target: aarch64-unknown-linux-gnu, tool: "linkchecker", path: "src/tools/linkchecker", mode: ToolBootstrap, source_type: InTree, extra_features: [], allow_features: "", cargo_args: [], artifact_kind: Binary } -- 0.137
[TIMING:end] tool::Linkchecker { compiler: Compiler { stage: 0, host: aarch64-unknown-linux-gnu, forced_compiler: false }, target: aarch64-unknown-linux-gnu } -- 0.000
##[group]Testing stage1 Linkcheck (aarch64-unknown-linux-gnu)
reference/names/name-resolution.html:452: broken link - `alloc/string/type.String.html`
reference/print.html:48602: broken link - `alloc/string/type.String.html`
proc_macro/struct.Punct.html:31: broken link - `alloc/string/generic/struct.String.html`
proc_macro/struct.Literal.html:295: broken link - `alloc/string/type.String.html`
proc_macro/struct.Literal.html:321: broken link - `alloc/string/generic/struct.String.html`
proc_macro/struct.Span.html:23: broken link - `alloc/string/type.String.html`
proc_macro/struct.Span.html:36: broken link - `alloc/string/type.String.html`
proc_macro/struct.Span.html:42: broken link - `alloc/string/type.String.html`
proc_macro/struct.Span.html:44: broken link - `alloc/string/type.String.html`
proc_macro/struct.Span.html:46: broken link - `alloc/string/type.String.html`
proc_macro/struct.Span.html:48: broken link - `alloc/string/type.String.html`
proc_macro/tracked/fn.env_var.html:1: broken link - `alloc/string/type.String.html`
proc_macro/struct.Ident.html:39: broken link - `alloc/string/generic/struct.String.html`
proc_macro/struct.ExpandError.html:13: broken link - `alloc/string/generic/struct.String.html`
proc_macro/struct.Diagnostic.html:3: broken link - `alloc/string/type.String.html`
proc_macro/struct.Diagnostic.html:6: broken link - `alloc/string/type.String.html`
proc_macro/struct.Diagnostic.html:10: broken link - `alloc/string/type.String.html`
proc_macro/struct.Diagnostic.html:11: broken link - `alloc/string/type.String.html`
proc_macro/struct.Diagnostic.html:14: broken link - `alloc/string/type.String.html`
proc_macro/struct.Diagnostic.html:15: broken link - `alloc/string/type.String.html`
proc_macro/struct.Diagnostic.html:18: broken link - `alloc/string/type.String.html`
proc_macro/struct.Diagnostic.html:19: broken link - `alloc/string/type.String.html`
proc_macro/struct.Diagnostic.html:22: broken link - `alloc/string/type.String.html`
proc_macro/struct.Diagnostic.html:23: broken link - `alloc/string/type.String.html`
proc_macro/struct.Diagnostic.html:27: broken link - `alloc/string/type.String.html`
proc_macro/enum.TokenTree.html:41: broken link - `alloc/string/generic/struct.String.html`
proc_macro/struct.LexError.html:15: broken link - `alloc/string/generic/struct.String.html`
proc_macro/trait.ToTokens.html:43: broken link - `alloc/string/type.String.html`
proc_macro/struct.TokenStream.html:49: broken link - `alloc/string/generic/struct.String.html`
proc_macro/struct.Group.html:38: broken link - `alloc/string/generic/struct.String.html`
test/struct.TestDesc.html:14: broken link - `alloc/string/type.String.html`
test/enum.TestName.html:3: broken link - `alloc/string/type.String.html`
test/enum.TestName.html:5: broken link - `alloc/string/type.String.html`
test/enum.TestName.html:19: broken link - `alloc/string/generic/struct.String.html`
test/test/struct.TestDesc.html:14: broken link - `alloc/string/type.String.html`
test/test/enum.TestName.html:3: broken link - `alloc/string/type.String.html`
test/test/enum.TestName.html:5: broken link - `alloc/string/type.String.html`
test/test/enum.TestName.html:19: broken link - `alloc/string/generic/struct.String.html`
test/test/fn.parse_opts.html:1: broken link - `alloc/string/type.String.html`
test/test/fn.parse_opts.html:1: broken link - `alloc/string/type.String.html`
test/test/struct.MetricMap.html:12: broken link - `alloc/string/type.String.html`
test/test/struct.TestExecTime.html:14: broken link - `alloc/string/generic/struct.String.html`
test/test/struct.TestOpts.html:3: broken link - `alloc/string/type.String.html`
test/test/struct.TestOpts.html:17: broken link - `alloc/string/type.String.html`
test/test/struct.TestOpts.html:21: broken link - `alloc/string/type.String.html`
test/test/struct.TestOpts.html:21: broken link - `alloc/string/type.String.html`
test/test/enum.TestResult.html:4: broken link - `alloc/string/type.String.html`
test/test/enum.TestResult.html:8: broken link - `alloc/string/type.String.html`
test/struct.TestOpts.html:3: broken link - `alloc/string/type.String.html`
test/struct.TestOpts.html:17: broken link - `alloc/string/type.String.html`
test/struct.TestOpts.html:21: broken link - `alloc/string/type.String.html`
test/struct.TestOpts.html:21: broken link - `alloc/string/type.String.html`
test/fn.print_merged_doctests_times.html:2: broken link - `alloc/string/type.String.html`
test/fn.test_main.html:2: broken link - `alloc/string/type.String.html`
test/bench/struct.Bencher.html:10: broken link - `alloc/string/type.String.html`
test/bench/struct.Bencher.html:11: broken link - `alloc/string/type.String.html`
test/bench/fn.run_once.html:1: broken link - `alloc/string/type.String.html`
test/bench/fn.run_once.html:2: broken link - `alloc/string/type.String.html`
test/bench/fn.fmt_bench_samples.html:1: broken link - `alloc/string/type.String.html`
test/bench/fn.benchmark.html:8: broken link - `alloc/string/type.String.html`
test/fn.test_main_with_exit_callback.html:2: broken link - `alloc/string/type.String.html`
test/enum.TestFn.html:2: broken link - `alloc/string/type.String.html`
test/enum.TestFn.html:3: broken link - `alloc/string/type.String.html`
test/enum.TestFn.html:4: broken link - `alloc/string/type.String.html`
test/enum.TestFn.html:5: broken link - `alloc/string/type.String.html`
test/enum.TestFn.html:6: broken link - `alloc/string/type.String.html`
test/enum.TestFn.html:7: broken link - `alloc/string/type.String.html`
test/enum.TestFn.html:8: broken link - `alloc/string/type.String.html`
test/enum.TestFn.html:8: broken link - `alloc/string/type.String.html`
test/enum.TestFn.html:8: broken link - `alloc/string/type.String.html`
test/enum.TestFn.html:8: broken link - `alloc/string/type.String.html`
test/enum.TestFn.html:8: broken link - `alloc/string/type.String.html`
test/enum.TestFn.html:8: broken link - `alloc/string/type.String.html`
test/fn.assert_test_result.html:1: broken link - `alloc/string/type.String.html`
std/ops/trait.Deref.html:61: broken link - `alloc/string/type.String.html`
core/ops/trait.Deref.html:61: broken link - `alloc/string/type.String.html`
book/ch02-00-guessing-game-tutorial.html:366: broken link - `std/string/struct.String.html`
book/print.html:1078: broken link - `std/string/struct.String.html`
checked links in: 26.2s
number of HTML files scanned: 62152
number of HTML redirects found: 22160
number of links checked: 2912997
number of links ignored due to external: 179110
number of links ignored due to exceptions: 24
number of intra doc links ignored: 9
errors found: 78
found some broken links
NOTE: if you are adding or renaming a markdown file in a mdBook, don't forget to register the page in SUMMARY.md
Bootstrap failed while executing `--stage 2 test --skip tidy --skip tests --skip coverage-map --skip coverage-run --skip library --skip tidyselftest`
Command `/checkout/obj/build/aarch64-unknown-linux-gnu/stage1-tools-bin/linkchecker /checkout/obj/build/aarch64-unknown-linux-gnu/doc` failed with exit code 1
Created at: src/bootstrap/src/core/build_steps/tool.rs:1618:23
Executed at: src/bootstrap/src/core/build_steps/test.rs:191:77

---
   2: <std::backtrace::Backtrace>::create
             at /rustc/ef0fb8a2563200e322fa4419f09f65a63742038c/library/std/src/backtrace.rs:331:13
   3: <bootstrap::utils::exec::DeferredCommand>::finish_process
             at /checkout/src/bootstrap/src/utils/exec.rs:939:17
   4: <bootstrap::utils::exec::DeferredCommand>::wait_for_output::<&bootstrap::utils::exec::ExecutionContext>
             at /checkout/src/bootstrap/src/utils/exec.rs:831:21
   5: <bootstrap::utils::exec::ExecutionContext>::run
             at /checkout/src/bootstrap/src/utils/exec.rs:741:45
   6: <bootstrap::utils::exec::BootstrapCommand>::run::<&bootstrap::core::builder::Builder>
             at /checkout/src/bootstrap/src/utils/exec.rs:339:27
   7: <bootstrap::core::build_steps::test::Linkcheck as bootstrap::core::builder::Step>::run
             at /checkout/src/bootstrap/src/core/build_steps/test.rs:191:77
   8: <bootstrap::core::builder::Builder>::ensure::<bootstrap::core::build_steps::test::Linkcheck>
             at /checkout/src/bootstrap/src/core/builder/mod.rs:1595:36
   9: <bootstrap::core::build_steps::test::Linkcheck as bootstrap::core::builder::Step>::make_run
             at /checkout/src/bootstrap/src/core/build_steps/test.rs:137:21
  10: <bootstrap::core::builder::StepDescription>::maybe_run
             at /checkout/src/bootstrap/src/core/builder/mod.rs:476:13
  11: bootstrap::core::builder::cli_paths::match_paths_to_steps_and_run
             at /checkout/src/bootstrap/src/core/builder/cli_paths.rs:141:22
  12: <bootstrap::core::builder::Builder>::run_step_descriptions
             at /checkout/src/bootstrap/src/core/builder/mod.rs:1138:9
  13: <bootstrap::core::builder::Builder>::execute_cli
             at /checkout/src/bootstrap/src/core/builder/mod.rs:1117:14
  14: <bootstrap::Build>::build
             at /checkout/src/bootstrap/src/lib.rs:803:25
  15: bootstrap::main
             at /checkout/src/bootstrap/src/bin/main.rs:130:11
  16: <fn() as core::ops::function::FnOnce<()>>::call_once
             at /rustc/ef0fb8a2563200e322fa4419f09f65a63742038c/library/core/src/ops/function.rs:250:5
  17: std::sys::backtrace::__rust_begin_short_backtrace::<fn(), ()>
             at /rustc/ef0fb8a2563200e322fa4419f09f65a63742038c/library/std/src/sys/backtrace.rs:166:18
  18: std::rt::lang_start::<()>::{closure#0}
             at /rustc/ef0fb8a2563200e322fa4419f09f65a63742038c/library/std/src/rt.rs:206:18
  19: <&dyn core::ops::function::Fn<(), Output = i32> + core::panic::unwind_safe::RefUnwindSafe + core::marker::Sync as core::ops::function::FnOnce<()>>::call_once
             at /rustc/ef0fb8a2563200e322fa4419f09f65a63742038c/library/core/src/ops/function.rs:287:21
  20: std::panicking::catch_unwind::do_call::<&dyn core::ops::function::Fn<(), Output = i32> + core::panic::unwind_safe::RefUnwindSafe + core::marker::Sync, i32>
             at /rustc/ef0fb8a2563200e322fa4419f09f65a63742038c/library/std/src/panicking.rs:581:40
  21: std::panicking::catch_unwind::<i32, &dyn core::ops::function::Fn<(), Output = i32> + core::panic::unwind_safe::RefUnwindSafe + core::marker::Sync>
             at /rustc/ef0fb8a2563200e322fa4419f09f65a63742038c/library/std/src/panicking.rs:544:19
  22: std::panic::catch_unwind::<&dyn core::ops::function::Fn<(), Output = i32> + core::panic::unwind_safe::RefUnwindSafe + core::marker::Sync, i32>
             at /rustc/ef0fb8a2563200e322fa4419f09f65a63742038c/library/std/src/panic.rs:359:14
  23: std::rt::lang_start_internal::{closure#0}
             at /rustc/ef0fb8a2563200e322fa4419f09f65a63742038c/library/std/src/rt.rs:175:24
  24: std::panicking::catch_unwind::do_call::<std::rt::lang_start_internal::{closure#0}, isize>
             at /rustc/ef0fb8a2563200e322fa4419f09f65a63742038c/library/std/src/panicking.rs:581:40
---
  31: __libc_start_main
  32: _start


Command has failed. Rerun with -v to see more details.
Build completed unsuccessfully in 0:40:12
  local time: Sat May  9 22:32:51 UTC 2026
  network time: Sat, 09 May 2026 22:32:51 GMT
##[error]Process completed with exit code 1.
##[group]Run echo "disk usage:"
```

</details>

shua committed:
shenanigans in the prelude

want to see if I can slip this `String<Global>` typedef into the prelude,
and see maybe a crater-run shows we can get away with not figuring out
the type inference + struct with single default generic param
ambiguity.

This change has another improvement over the previous, in that *most*
debug displays of the `String` are just `std::string::String`, where
the previous would resolve to `std::string::generic::String`.
Also the docs are all in the right place, and are for `struct.String.html`
rather than `type.String.html`

rust-log-analyzer commented:
The job **`tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/28190555663/job/83503986857) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/83503986857) [(plain)](https://github.com/rust-lang/rust/commit/5868fe7a8bc6cde721062367699dddf184bcc1cf/checks/83503986857/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
fmt: checked 6938 files
tidy check
tidy [rustdoc_json (src)]: `rustdoc-json-types` modified, checking format version
tidy: Skipping binary file check, read-only filesystem
tidy [features]: /checkout/library/std/src/prelude/mod.rs:193: feature gate allocator_api has inconsistent `issue`: "none" mismatches the previous `issue` of "32838"
tidy [features]: /checkout/library/std/src/collections/hash/set.rs:129: feature gate allocator_api has inconsistent `issue`: "32838" mismatches the previous `issue` of "none"
tidy [features]: FAIL
removing old virtual environment
creating virtual environment at '/checkout/obj/build/venv' using 'python3.10' and 'venv'
creating virtual environment at '/checkout/obj/build/venv' using 'python3.10' and 'virtualenv'
Requirement already satisfied: pip in ./build/venv/lib/python3.10/site-packages (26.1.1)
Collecting pip
---
Running eslint on rustdoc JS files
info: ES-Check: checking 7 files...
info: ✓ ES-Check passed! All files are ES10 compatible.
typechecking javascript files
tidy: The following check failed: features
Command `/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools-bin/rust-tidy --root-path=/checkout --cargo-path=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo --output-dir=/checkout/obj/build --concurrency=4 --npm-path=/node/bin/yarn --ci=true --extra-checks=py,cpp,js,spellcheck` failed with exit code 1
Created at: src/bootstrap/src/core/build_steps/tool.rs:1625:23
Executed at: src/bootstrap/src/core/build_steps/test.rs:1645:29

--- BACKTRACE vvv
   0: <bootstrap::utils::exec::DeferredCommand>::finish_process
             at /checkout/src/bootstrap/src/utils/exec.rs:939:17
   1: <bootstrap::utils::exec::DeferredCommand>::wait_for_output::<&bootstrap::utils::exec::ExecutionContext>
             at /checkout/src/bootstrap/src/utils/exec.rs:831:21
   2: <bootstrap::utils::exec::ExecutionContext>::run
             at /checkout/src/bootstrap/src/utils/exec.rs:741:45
   3: <bootstrap::utils::exec::BootstrapCommand>::run::<&bootstrap::core::builder::Builder>
             at /checkout/src/bootstrap/src/utils/exec.rs:339:27
   4: <bootstrap::core::build_steps::test::Tidy as bootstrap::core::builder::Step>::run
             at /checkout/src/bootstrap/src/core/build_steps/test.rs:1645:29
   5: <bootstrap::core::builder::Builder>::ensure::<bootstrap::core::build_steps::test::Tidy>
             at /checkout/src/bootstrap/src/core/builder/mod.rs:1597:36
   6: <bootstrap::core::build_steps::test::Tidy as bootstrap::core::builder::Step>::make_run
             at /checkout/src/bootstrap/src/core/build_steps/test.rs:1567:21
   7: <bootstrap::core::builder::StepDescription>::maybe_run
             at /checkout/src/bootstrap/src/core/builder/mod.rs:476:13
   8: bootstrap::core::builder::cli_paths::match_paths_to_steps_and_run
             at /checkout/src/bootstrap/src/core/builder/cli_paths.rs:232:18
   9: <bootstrap::core::builder::Builder>::run_step_descriptions
             at /checkout/src/bootstrap/src/core/builder/mod.rs:1140:9
  10: <bootstrap::core::builder::Builder>::execute_cli
             at /checkout/src/bootstrap/src/core/builder/mod.rs:1119:14
  11: <bootstrap::Build>::build
             at /checkout/src/bootstrap/src/lib.rs:803:25
  12: bootstrap::main
             at /checkout/src/bootstrap/src/bin/main.rs:130:11
  13: <fn() as core::ops::function::FnOnce<()>>::call_once
             at /rustc/0417c25868d6dfbd1c291dfeae950504faa6f790/library/core/src/ops/function.rs:250:5
  14: std::sys::backtrace::__rust_begin_short_backtrace::<fn(), ()>
             at /rustc/0417c25868d6dfbd1c291dfeae950504faa6f790/library/std/src/sys/backtrace.rs:166:18
  15: std::rt::lang_start::<()>::{closure#0}
             at /rustc/0417c25868d6dfbd1c291dfeae950504faa6f790/library/std/src/rt.rs:206:18
  16: <&dyn core::ops::function::Fn<(), Output = i32> + core::marker::Sync + core::panic::unwind_safe::RefUnwindSafe as core::ops::function::FnOnce<()>>::call_once
             at /rustc/0417c25868d6dfbd1c291dfeae950504faa6f790/library/core/src/ops/function.rs:287:21
  17: std::panicking::catch_unwind::do_call::<&dyn core::ops::function::Fn<(), Output = i32> + core::marker::Sync + core::panic::unwind_safe::RefUnwindSafe, i32>
             at /rustc/0417c25868d6dfbd1c291dfeae950504faa6f790/library/std/src/panicking.rs:581:40
  18: std::panicking::catch_unwind::<i32, &dyn core::ops::function::Fn<(), Output = i32> + core::marker::Sync + core::panic::unwind_safe::RefUnwindSafe>
             at /rustc/0417c25868d6dfbd1c291dfeae950504faa6f790/library/std/src/panicking.rs:544:19
  19: std::panic::catch_unwind::<&dyn core::ops::function::Fn<(), Output = i32> + core::marker::Sync + core::panic::unwind_safe::RefUnwindSafe, i32>
             at /rustc/0417c25868d6dfbd1c291dfeae950504faa6f790/library/std/src/panic.rs:359:14
  20: std::rt::lang_start_internal::{closure#0}
             at /rustc/0417c25868d6dfbd1c291dfeae950504faa6f790/library/std/src/rt.rs:175:24
  21: std::panicking::catch_unwind::do_call::<std::rt::lang_start_internal::{closure#0}, isize>
             at /rustc/0417c25868d6dfbd1c291dfeae950504faa6f790/library/std/src/panicking.rs:581:40
---
  28: __libc_start_main
  29: _start


Command has failed. Rerun with -v to see more details.
Bootstrap failed while executing `test src/tools/tidy tidyselftest --extra-checks=py,cpp,js,spellcheck`
Build completed unsuccessfully in 0:02:46
  local time: Thu Jun 25 18:14:34 UTC 2026
  network time: Thu, 25 Jun 2026 18:14:34 GMT
##[error]Process completed with exit code 1.
```

</details>

rust-log-analyzer commented:
The job **`pr-check-2`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/28321793895/job/83904939410) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/83904939410) [(plain)](https://github.com/rust-lang/rust/commit/6aca0b996689ba158e3043c02e30fc8c8b3fe5db/checks/83904939410/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

```

</details>
